zomes/chat/src/batching_helper.rs

```diff
@@ -42,44 +42,54 @@ pub fn get_message_links(
     let mut depth = 0;
     while links.len() < target_count && current_search_path.as_ref().len() >= root_path_length {
         if current_search_path.exists()? {
-            // let earliest_seen_child_segment =
-            //     last_segment_from_path(&earliest_seen_child_path).unwrap();
+            let earliest_seen_child_segment =
+                last_segment_from_path(&earliest_seen_child_path).unwrap();
             let children = current_search_path.children()?;
-            // This was just an optimization which we are going to loose
-            // children.retain(|child_link| {
-            //     link_is_earlier(child_link, earliest_seen_child_segment).unwrap_or(false)
-            // });
+            debug!(
+                "current_search_path {:?} depth {} children {:?}",
+                current_search_path,
+                depth,
+                children
+                    .iter()
+                    .map(|l| (&l.tag, l.timestamp))
+                    .collect::<Vec<_>>()
+            );
+            let mut children = children
+                .into_iter()
+                .filter_map(|l| path_component_from_link(&l).ok().map(|c| (c, l))) // filter out non-path links
+                .map(|(c, l)| Ok((segment_from_component(&c)?, l)))
+                .collect::<Result<Vec<_>, ChatError>>()?;
+
+            children.retain(|(segment, _)| *segment < earliest_seen_child_segment);
             append_message_links_recursive(children, &mut links, target_count, depth)?;
         }
 
         earliest_seen_child_path = current_search_path;
         current_search_path = earliest_seen_child_path.parent().unwrap();
         depth += 1;
     }
-    let mut clean_links_list = Vec::new();
-    links.iter().for_each(|l| {
-        if !clean_links_list.iter().any(|ll| ll == l) {
-            clean_links_list.push(l.clone())
-        }
-    });
 
-    Ok(clean_links_list)
+    Ok(links)
 }
 
 fn append_message_links_recursive(
-    mut children: Vec<Link>,
+    mut children: Vec<(i32, Link)>,
     links: &mut Vec<Link>,
     target_count: usize,
     depth: u8,
 ) -> ChatResult<()> {
-    children.sort_unstable_by_key(|grandchild_link| cmp::Reverse(grandchild_link.timestamp));
-    for child_link in children {
+    children.sort_unstable_by_key(|(segment, _)| cmp::Reverse(*segment));
```

Comment:
It was important to sort by segment instead of timestamp, since in the new test, fake messages are inserted with chosen time-path-segments, but the timestamp is not fake and still represents the system time